### DISCUSSION TITLE
riverpod within a flutter package

### DISCUSSION THREAD

**Discussion by neiljaywarner** [rrousselGit/riverpod#1613] [Q&A]
Hi!  I'm trying to use riverpod for a completely self-contained package.

is this a dumb idea?  It seemed to be going fairly well or even really well until we navigated to another page.  Then as you can imagine I lost the providerscope since the provider scope was in the "inner shell" module.

eg
Outer Dummy App1 `runApp(DriverWidgetThatHasAppBarAndScaffoldAndHardcodedParametersAndWrapsActualPackageWrapperWidget())`
Real OuterApp: tap some things, get some info from db/network/whatever, pass it to ActualPackageWrapperWidget(param1,param2)

package widget is in directory project root/pakcage/actual_package and has its own pubspec.yaml obviously etc

ActualPackageWrapperWidget is like this
`
class ActualPackageWrapperWidget extends StatelessWidget {
  const ActualPackageWrapperWidget({Key? key, required this.param1, required this.param2}) : super(key: key);
  final String param1, param2;

  @override
  Widget build(BuildContext context) => ProviderScope(child: ActualPackageWidget(param1,param2));
}
`
so as you can imagine since MaterialApp is in the "host" package then pushing a new navigator route loses my ProviderScope.

Of course, the biggest thing this causes problems for is testing.
Possible solutions
1) get riverpod in the host app and make the testing in the host modules - this is not awesome because there's a fair amount of tests already, and the host app is by another team etc
2) MaterialApp within Material App - is this a terrible idea?
3) some way to shove the overrides of a providerscope into the next providerscope?

**Comment by rrousselGit** [comment]
You probably want to offer a widget that users should put above their app

**Comment by neiljaywarner** [comment]
I want a flutter package that can go inside an Android iOS flutter or web
app.

It just happens to be going in a flutter app so far



....sent from my phone

**Comment by akshaydoshi2** [comment]
@neiljaywarner did you get a solution?
I am trying to use Riverpod in a flutter package as well. But I'm not sure how `ProviderScope` should be handled.
I don't think MaterialApp within MaterialApp is a good idea. And I don't want to add Riverpod as a dependency in the parent app.

**Comment by neiljaywarner** [comment]
I don't recall but try remis idea :)
Let me know if it doesn't work and I'll check. I think it worked out fine eventually

**Reply by akshaydoshi2** [reply]
@neiljaywarner Can you give me a link to  remis idea?
I can’t seem to find it

**Comment by neiljaywarner** [comment]
https://github.com/rrousselGit/riverpod/discussions/1613#discussioncomment-3613311

**Reply by akshaydoshi2** [reply]
Ah Thanks!
But I'm unable understand what exactly to do to `offer a widget that users should put above their app`.
Can you help me with that?

**Comment by neiljaywarner** [comment]
What's the purpose of your package and how does the user interact with it?

If they trigger it with a button then either the screen that launches or
the button could be the widget for example

....sent from my phone

**Reply by akshaydoshi2** [reply]
My package has 3 widgets (3 screens to be precise). The user can interact with the widget on a button press and can simply `navigator.push` to the package.
The user can pop these screens and come back to the parent app.

Right now I have to wrap the `main()` in the parent app with a ProviderScope for everything to function.

What I want to do is make the package independent so that the user does not have to import riverpod and wrap it with ProviderScope in the parent app.

**Comment by neiljaywarner** [comment]
So wrap the screen you navigate to in providerscope?


....sent from my phone

**Reply by akshaydoshi2** [reply]
Yeah I came up with the same solution. But wouldn't that create multiple instances of the providerscope/providercontainer?

**Reply by akshaydoshi2** [reply]
@neiljaywarner I did try your suggestion. But it doesn’t seem to work. It creates separate instances of ProviderScope and so separate instances of the providers are also created.

**Comment by neiljaywarner** [comment]
Maybe put both buttons in a container or stack? :)

....sent from my phone

**Comment by jordanb** [comment]
Probably too late but the solution is to encapsulate your "pages" inside your own navigator

```dart
final _navigatorKey = GlobalKey<NavigatorState>();

@override
Widget build(BuildContext context) {
  return ProviderScope(
    child: Navigator(
      key: _navigatorKey,
      initialRoute: 'exampleRouteName',
      onGenerateRoute: _onGenerateRoute,
    ),
  );
}
```

then use `_navigatorKey.currentState!.pushNamed(<name here>)`
